# 결정 트리 실습

사용자 행동 인식 데이터 세트

[원본 데이터셋]
https://archive.ics.uci.edu/ml/datasets/human+activity+recognition+using+smartphones

데이터 세트 정보:

실험은 19-48세 사이의 30명의 지원자 그룹으로 수행되었습니다. 각자 스마트폰(Samsung Galaxy S II)을 허리에 차고 6가지 활동(
1:WALKING, 
2:WALKING_UPSTAIRS, 
3:WALKING_DOWNSTAIRS, 4:SITTING, 5:STANDING, 6:LAYING)을 수행했습니다. 내장된 가속도계와 자이로스코프를 사용하여 50Hz의 일정한 속도로 3축 선형 가속도와 3축 각속도를 캡처했습니다. 실험은 데이터에 수동으로 레이블을 지정하기 위해 비디오로 녹화되었습니다. 얻은 데이터 세트는 훈련 데이터 생성을 위해 70%, 테스트 데이터 생성을 위해 30%가 선택된 두 세트로 무작위로 분할되었습니다.

센서 신호(가속도계 및 자이로스코프)는 노이즈 필터를 적용하여 사전 처리된 다음 2.56초 및 50% 중첩(128 판독/창)의 고정 너비 슬라이딩 창에서 샘플링되었습니다. 중력 및 신체 운동 성분을 갖는 센서 가속도 신호는 Butterworth 저역 통과 필터를 사용하여 신체 가속도와 중력으로 분리되었습니다. 중력은 저주파 성분만 있다고 가정하므로 차단 주파수가 0.3Hz인 필터를 사용했습니다. 각 창에서 시간 및 주파수 영역에서 변수를 계산하여 특징 벡터를 얻었습니다.

이 데이터 세트에 대한 자세한 내용은 README.txt 파일을 확인하십시오.

참가자 중 한 명과 함께 녹화된 6가지 활동의 예를 포함하는 실험 비디오는 다음 링크에서 볼 수 있습니다.

https://www.youtube.com/watch?v=XOEN9W05_4A]


**Mission**

수집된 데이터를 기반으로 결정트리를 이용해 어떤 행동을 하고 있는지 예측해 봅시다.

In [1]:
import pandas as pd
import matplotlib.pyplot as plt

## 데이터 로드

1. 피처만 있는 파일을 로드
2. 훈련데이터와 피처를 합쳐 훈련 데이터 프레임 만듬
3. 데스트데이터와 피처를 합쳐 테스트 데이터 프레임 만듬

In [2]:
## 데이터 로드하는 거요~~~

import pandas as pd

# 피처이름 로드
feature_name_df = pd.read_csv('./dataset/human_activity/features_new.txt', header=None,names=['column_name'])

# 피처이름을 리스트로
feature_name = feature_name_df['column_name'].values.tolist()

# 학습 피처 데이터 셋과 테스트 피처 데이터을 DataFrame으로 로딩. 컬럼명은 feature_name 적용
X_train = pd.read_csv('./dataset/human_activity/train/X_train.txt',sep='\s+', names=feature_name )
X_test = pd.read_csv('./dataset/human_activity/test/X_test.txt',sep='\s+', names=feature_name)

# 학습 레이블과 테스트 레이블 데이터을 DataFrame으로 로딩하고 컬럼명은 action으로 부여
y_train = pd.read_csv('./dataset/human_activity/train/y_train.txt',sep='\s+',header=None,names=['action'])
y_test = pd.read_csv('./dataset/human_activity/test/y_test.txt',sep='\s+',header=None,names=['action'])

In [3]:
# 피처이름 로드 
feature_name_df = pd.read_csv('./dataset/human_activity/features_new.txt', header=None,names=['column_name'])


In [4]:
feature_name_df

column_name
0                       tBodyAcc-mean()-X
1                       tBodyAcc-mean()-Y
2                       tBodyAcc-mean()-Z
3                        tBodyAcc-std()-X
4                        tBodyAcc-std()-Y
..                                    ...
556      angle(tBodyGyroMean,gravityMean)
557  angle(tBodyGyroJerkMean,gravityMean)
558                  angle(X,gravityMean)
559                  angle(Y,gravityMean)
560                  angle(Z,gravityMean)

[561 rows x 1 columns]

In [5]:
# 피처이름을 리스트로

feature_name = feature_name_df['column_name'].values.tolist()

In [6]:
feature_name

['tBodyAcc-mean()-X',
 'tBodyAcc-mean()-Y',
 'tBodyAcc-mean()-Z',
 'tBodyAcc-std()-X',
 'tBodyAcc-std()-Y',
 'tBodyAcc-std()-Z',
 'tBodyAcc-mad()-X',
 'tBodyAcc-mad()-Y',
 'tBodyAcc-mad()-Z',
 'tBodyAcc-max()-X',
 'tBodyAcc-max()-Y',
 'tBodyAcc-max()-Z',
 'tBodyAcc-min()-X',
 'tBodyAcc-min()-Y',
 'tBodyAcc-min()-Z',
 'tBodyAcc-sma()',
 'tBodyAcc-energy()-X',
 'tBodyAcc-energy()-Y',
 'tBodyAcc-energy()-Z',
 'tBodyAcc-iqr()-X',
 'tBodyAcc-iqr()-Y',
 'tBodyAcc-iqr()-Z',
 'tBodyAcc-entropy()-X',
 'tBodyAcc-entropy()-Y',
 'tBodyAcc-entropy()-Z',
 'tBodyAcc-arCoeff()-X,1',
 'tBodyAcc-arCoeff()-X,2',
 'tBodyAcc-arCoeff()-X,3',
 'tBodyAcc-arCoeff()-X,4',
 'tBodyAcc-arCoeff()-Y,1',
 'tBodyAcc-arCoeff()-Y,2',
 'tBodyAcc-arCoeff()-Y,3',
 'tBodyAcc-arCoeff()-Y,4',
 'tBodyAcc-arCoeff()-Z,1',
 'tBodyAcc-arCoeff()-Z,2',
 'tBodyAcc-arCoeff()-Z,3',
 'tBodyAcc-arCoeff()-Z,4',
 'tBodyAcc-correlation()-X,Y',
 'tBodyAcc-correlation()-X,Z',
 'tBodyAcc-correlation()-Y,Z',
 'tGravityAcc-mean()-X',
 'tGravityA

In [7]:
len(feature_name)

561

## 학습 피처 데이터 셋과 테스트 피처 데이터을 DataFrame으로 로딩. 컬럼명은 feature_name 적용

In [8]:
# X_train

with open('./dataset/human_activity/train/X_train.txt', 'r') as f:
    X_train = f.readlines()

X_train_split = [x.strip().split() for x in X_train[0:]]

X_train_df = pd.DataFrame(X_train_split, columns = feature_name)

In [9]:
X_train_df

tBodyAcc-mean()-X tBodyAcc-mean()-Y tBodyAcc-mean()-Z tBodyAcc-std()-X  \
0       2.8858451e-001   -2.0294171e-002   -1.3290514e-001  -9.9527860e-001   
1       2.7841883e-001   -1.6410568e-002   -1.2352019e-001  -9.9824528e-001   
2       2.7965306e-001   -1.9467156e-002   -1.1346169e-001  -9.9537956e-001   
3       2.7917394e-001   -2.6200646e-002   -1.2328257e-001  -9.9609149e-001   
4       2.7662877e-001   -1.6569655e-002   -1.1536185e-001  -9.9813862e-001   
...                ...               ...               ...              ...   
7347    2.9966534e-001   -5.7193414e-002   -1.8123302e-001  -1.9538652e-001   
7348    2.7385271e-001   -7.7493259e-003   -1.4746837e-001  -2.3530853e-001   
7349    2.7338737e-001   -1.7010616e-002   -4.5021828e-002  -2.1821818e-001   
7350    2.8965416e-001   -1.8843044e-002   -1.5828059e-001  -2.1913944e-001   
7351    3.5150347e-001   -1.2423118e-002   -2.0386717e-001  -2.6927044e-001   

     tBodyAcc-std()-Y tBodyAcc-std()-Z tBodyAcc-mad()-X tBodyAcc-mad()-Y  \
0     -9.8311061e-001  -9.1352645e-001  -9.9511208e-001  -9.8318457e-001   
1     -9.7530022e-001  -9.6032199e-001  -9.9880719e-001  -9.7491437e-001   
2     -9.6718701e-001  -9.7894396e-001  -9.9651994e-001  -9.6366837e-001   
3     -9.8340270e-001  -9.9067510e-001  -9.9709947e-001  -9.8274984e-001   
4     -9.8081727e-001  -9.9048163e-001  -9.9832113e-001  -9.7967187e-001   
...               ...              ...              ...              ...   
7347   3.9904850e-002   7.7078081e-002  -2.8230064e-001   4.3615631e-002   
7348   4.8162805e-003   5.9279994e-002  -3.2255234e-001  -2.9456250e-002   
7349  -1.0382198e-001   2.7453270e-001  -3.0451515e-001  -9.8913034e-002   
7350  -1.1141169e-001   2.6889320e-001  -3.1048749e-001  -6.8200325e-002   
7351  -8.7211541e-002   1.7740393e-001  -3.7740405e-001  -3.8678063e-002   

     tBodyAcc-mad()-Z tBodyAcc-max()-X  ... fBodyBodyGyroJerkMag-meanFreq()  \
0     -9.2352702e-001  -9.3472378e-001  ...                 -7.4323027e-002   
1     -9.5768622e-001  -9.4306751e-001  ...                  1.5807454e-001   
2     -9.7746859e-001  -9.3869155e-001  ...                  4.1450281e-001   
3     -9.8930250e-001  -9.3869155e-001  ...                  4.0457253e-001   
4     -9.9044113e-001  -9.4246912e-001  ...                  8.7753013e-002   
...               ...              ...  ...                             ...   
7347   6.0410083e-002   2.1079544e-001  ...                 -7.0156695e-002   
7348   8.0585116e-002   1.1744028e-001  ...                  1.6525919e-001   
7349   3.3258449e-001   4.3998772e-002  ...                  1.9503401e-001   
7350   3.1947326e-001   1.0170184e-001  ...                  1.3865423e-002   
7351   2.2942952e-001   2.6901269e-001  ...                 -5.8401607e-002   

     fBodyBodyGyroJerkMag-skewness() fBodyBodyGyroJerkMag-kurtosis()  \
0                    -2.9867637e-001                 -7.1030407e-001   
1                    -5.9505094e-001                 -8.6149931e-001   
2                    -3.9074815e-001                 -7.6010372e-001   
3                    -1.1729020e-001                 -4.8284451e-001   
4                    -3.5147093e-001                 -6.9920515e-001   
...                              ...                             ...   
7347                 -5.8843274e-001                 -8.8032443e-001   
7348                 -3.9073832e-001                 -6.8074445e-001   
7349                  2.5145333e-002                 -3.0402936e-001   
7350                  6.3906883e-002                 -3.4431361e-001   
7351                 -3.8705200e-001                 -7.4073820e-001   

     angle(tBodyAccMean,gravity) angle(tBodyAccJerkMean),gravityMean)  \
0                -1.1275434e-001                       3.0400372e-002   
1                 5.3476955e-002                      -7.4345661e-003   
2                -1.1855926e-001                       1.7789948e-001   
3                -3.6787973e-002     

In [10]:
# X_test

with open('./dataset/human_activity/test/X_test.txt', 'r') as f:
    X_test = f.readlines()

X_test_split = [x.strip().split() for x in X_test[0:]]

X_test_df = pd.DataFrame(X_test_split, columns = feature_name)

## 학습 레이블과 테스트 레이블 데이터을 DataFrame으로 로딩하고 컬럼명은 action으로 부여

In [11]:
# y_train

y_train = pd.read_csv('./dataset/human_activity/train/y_train.txt', header=None,names=['action'])
y_train_df = pd.DataFrame(data= y_train)


In [12]:
y_train_df

action
0          5
1          5
2          5
3          5
4          5
...      ...
7347       2
7348       2
7349       2
7350       2
7351       2

[7352 rows x 1 columns]

In [13]:
# 피처이름을 리스트로
# 피처이름 로드 
activity_labels_df = pd.read_csv('./dataset/human_activity/activity_labels.txt', header=None,names=['column_name'])

activity_labels = activity_labels_df['column_name'].values.tolist()

In [14]:
activity_labels

['1 WALKING',
 '2 WALKING_UPSTAIRS',
 '3 WALKING_DOWNSTAIRS',
 '4 SITTING',
 '5 STANDING',
 '6 LAYING']

In [15]:
# y_test

y_test = pd.read_csv('./dataset/human_activity/test/y_test.txt', header=None,names=['action'])
y_test_df = pd.DataFrame(data= y_test)


In [16]:
y_test_df

action
0          5
1          5
2          5
3          5
4          5
...      ...
2942       2
2943       2
2944       2
2945       2
2946       2

[2947 rows x 1 columns]

### action 피처의 class별 갯수


In [17]:
y_train_df.value_counts()

action
6         1407
5         1374
4         1286
1         1226
2         1073
3          986
dtype: int64

In [18]:
y_test.value_counts()

action
6         537
5         532
1         496
4         491
2         471
3         420
dtype: int64

In [19]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# DesisionTree분류 클래스 사용(randomstate 156)
dt_clf = DecisionTreeClassifier(random_state=156)


# 훈련
dt_clf.fit(X_train_df, y_train_df)

# 예측
pred = dt_clf.predict(X_test_df)


#평가
accuracy_score(y_test_df,pred)


0.8547675602307431

In [20]:
# DecisionTreeClassifier의 하이퍼 파라미터 추출
print('DecisionTreeClassifier 기본 하이퍼 파라미터:\n', dt_clf.get_params())

DecisionTreeClassifier 기본 하이퍼 파라미터:
 {'ccp_alpha': 0.0, 'class_weight': None, 'criterion': 'gini', 'max_depth': None, 'max_features': None, 'max_leaf_nodes': None, 'min_impurity_decrease': 0.0, 'min_samples_leaf': 1, 'min_samples_split': 2, 'min_weight_fraction_leaf': 0.0, 'random_state': 156, 'splitter': 'best'}


## 하이퍼파라미터 튜닝1

In [21]:
from sklearn.model_selection import GridSearchCV

#max_depth를 바꿔가며 훈련해보자
#나무의 깊이를 얼마만큼 성장시킬것인가

params = {
    'max_depth' : [ 6, 8 ,10, 12, 16 ,20, 24]
}


In [22]:
# GridSearchCV 사용


# 훈련
grid_clf = GridSearchCV(dt_clf, param_grid=params, scoring='accuracy', cv=3)
grid_clf.fit(X_train_df, y_train_df)


GridSearchCV(cv=3, estimator=DecisionTreeClassifier(random_state=156),
             param_grid={'max_depth': [6, 8, 10, 12, 16, 20, 24]},
             scoring='accuracy')

In [23]:
print('최적 하이퍼 파라미터:{0}, 최적 평균 정확도:{1:.3f}'.format(grid_clf.best_params_, 
                                                  grid_clf.best_score_))

최적 하이퍼 파라미터:{'max_depth': 6}, 최적 평균 정확도:0.847


In [24]:
# GridSearchCV객체의 cv_results_ 속성을 DataFrame으로 생성. 


In [25]:
#베스트 분류기 선택
best_clf = grid_clf.best_estimator_

#예측
pred = best_clf.predict(X_test_df)

#평가
accuracy_score(y_test_df,pred)


0.8557855446216491

```
    하이퍼파라미터 튜닝으로 성능이 향상 되었나요?
```

## 하이퍼파라미터 튜닝2

In [26]:
#하이퍼파라미터
params = {
    'max_depth' : [ 8 , 12, 16 ,20], 
    'min_samples_split' : [16,24],
}


In [27]:
# GridSearchCV 사용


# 훈련
grid_clf = GridSearchCV(dt_clf, param_grid=params, scoring='accuracy', cv=3)
grid_clf.fit(X_train_df, y_train_df)

print('최적 하이퍼 파라미터:{0}, 최적 평균 정확도:{1:.3f}'.format(grid_clf.best_params_, 
                                                  grid_clf.best_score_))


최적 하이퍼 파라미터:{'max_depth': 8, 'min_samples_split': 16}, 최적 평균 정확도:0.846


In [28]:
#베스트 분류기 선택
best_clf = grid_clf.best_estimator_

#예측
pred = best_clf.predict(X_test_df)

#평가
accuracy_score(y_test_df,pred)


0.8717339667458432

```
    하이퍼파라미터 튜닝으로 성능이 향상 되었나요?
```

* 0.855 에서 하이퍼 파라미터 튜닝1에서는 0.856으로 성능이 다소 향상되었고 하이퍼 파라미터 튜닝 2에서는 0.871으로 성능이 미세하게 향상되었다.

---

**[생활탐구]**


```
방문을 꼭 닫은채 공부하고 있다는 보검이...
정말 보검이는 공부하고 있을까요?
생체 신호 측정기를 착용한 보검이로 부터 아래와 같은 신호를 가져왔습니다. 
지금 현재 보검이는 무엇을 하고 있나요?

[보검이 생체 신호 파일]X_quiz.txt 

```

In [29]:
#신호 읽어서 예측해보기
# X_quiz

# 피처이름 로드 
X_quiz_df = pd.read_csv('./dataset/human_activity/test/X_quiz.txt', header=None, sep = ',')




In [30]:
X_quiz_df = X_quiz_df.T

In [31]:
X_quiz_df = X_quiz_df.drop([0], axis = 1)

In [32]:
X_quiz_df = X_quiz_df.T

In [33]:
X_quiz_df

0         1         2         3         4         5         6    \
1  0.257178 -0.023285 -0.014654 -0.938404 -0.920091 -0.667683 -0.952501   

        7         8         9    ...       551      552       553       554  \
1 -0.925249 -0.674302 -0.894088  ...  0.071645 -0.33037 -0.705974  0.006462   

       555       556       557       558       559       560  
1  0.16292 -0.825886  0.271151 -0.720009  0.276801 -0.057978  

[1 rows x 561 columns]

In [34]:
feature_name

['tBodyAcc-mean()-X',
 'tBodyAcc-mean()-Y',
 'tBodyAcc-mean()-Z',
 'tBodyAcc-std()-X',
 'tBodyAcc-std()-Y',
 'tBodyAcc-std()-Z',
 'tBodyAcc-mad()-X',
 'tBodyAcc-mad()-Y',
 'tBodyAcc-mad()-Z',
 'tBodyAcc-max()-X',
 'tBodyAcc-max()-Y',
 'tBodyAcc-max()-Z',
 'tBodyAcc-min()-X',
 'tBodyAcc-min()-Y',
 'tBodyAcc-min()-Z',
 'tBodyAcc-sma()',
 'tBodyAcc-energy()-X',
 'tBodyAcc-energy()-Y',
 'tBodyAcc-energy()-Z',
 'tBodyAcc-iqr()-X',
 'tBodyAcc-iqr()-Y',
 'tBodyAcc-iqr()-Z',
 'tBodyAcc-entropy()-X',
 'tBodyAcc-entropy()-Y',
 'tBodyAcc-entropy()-Z',
 'tBodyAcc-arCoeff()-X,1',
 'tBodyAcc-arCoeff()-X,2',
 'tBodyAcc-arCoeff()-X,3',
 'tBodyAcc-arCoeff()-X,4',
 'tBodyAcc-arCoeff()-Y,1',
 'tBodyAcc-arCoeff()-Y,2',
 'tBodyAcc-arCoeff()-Y,3',
 'tBodyAcc-arCoeff()-Y,4',
 'tBodyAcc-arCoeff()-Z,1',
 'tBodyAcc-arCoeff()-Z,2',
 'tBodyAcc-arCoeff()-Z,3',
 'tBodyAcc-arCoeff()-Z,4',
 'tBodyAcc-correlation()-X,Y',
 'tBodyAcc-correlation()-X,Z',
 'tBodyAcc-correlation()-Y,Z',
 'tGravityAcc-mean()-X',
 'tGravityA

In [35]:
X_quiz_df.columns = feature_name

In [36]:
X_quiz_df

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  tBodyAcc-std()-X  \
1           0.257178          -0.023285          -0.014654         -0.938404   

   tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  tBodyAcc-mad()-Y  \
1         -0.920091         -0.667683         -0.952501         -0.925249   

   tBodyAcc-mad()-Z  tBodyAcc-max()-X  ...  fBodyBodyGyroJerkMag-meanFreq()  \
1         -0.674302         -0.894088  ...                         0.071645   

   fBodyBodyGyroJerkMag-skewness()  fBodyBodyGyroJerkMag-kurtosis()  \
1                         -0.33037                        -0.705974   

   angle(tBodyAccMean,gravity)  angle(tBodyAccJerkMean),gravityMean)  \
1                     0.006462                               0.16292   

   angle(tBodyGyroMean,gravityMean)  angle(tBodyGyroJerkMean,gravityMean)  \
1                         -0.825886                              0.271151   

   angle(X,gravityMean)  angle(Y,gravityMean)  angle(Z,gravityMean)  
1             -0.720009              0.276801             -0.057978  

[1 rows x 561 columns]

In [37]:
best_clf.predict(X_quiz_df)

array([5], dtype=int64)

* 5번은 Standing. 보검이는 현재 일어서있다.
* 다른 사람들은 6번 LAYING이 나오는데, X_quizz.txt 파일의 값이 나와는 다르다.